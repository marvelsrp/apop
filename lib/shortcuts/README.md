<a name="shortcuts"></a>

## shortcuts : <code>object</code>
**Kind**: global namespace  

* [shortcuts](#shortcuts) : <code>object</code>
    * [.module.exports(origin)](#shortcuts.module.exports) ⇒ <code>function</code>
    * [.module.exports(origin, config)](#shortcuts.module.exports) ⇒ <code>function</code>
    * [.module.exports(origin)](#shortcuts.module.exports) ⇒ <code>function</code>
    * [.module.exports(origin)](#shortcuts.module.exports) ⇒ <code>function</code>
    * [.module.exports(origin, config)](#shortcuts.module.exports) ⇒ <code>function</code>

<a name="shortcuts.module.exports"></a>

### shortcuts.module.exports(origin) ⇒ <code>function</code>
Shortcut for clear object

**Kind**: static method of <code>[shortcuts](#shortcuts)</code>  
**Returns**: <code>function</code> - Returns transformed object with snake_case keys  
**See**: [Original method](../actions#actions.clear)  

| Param | Type | Description |
| --- | --- | --- |
| origin | <code>Object</code> | Origin object |

**Example**  
```js
import clear from 'rmk/shortcuts/clear';

let data = {foo:null, bar: 0, baz:null}
clear(data)
// => {bar:0}
```
<a name="shortcuts.module.exports"></a>

### shortcuts.module.exports(origin, config) ⇒ <code>function</code>
Rename state by config

**Kind**: static method of <code>[shortcuts](#shortcuts)</code>  
**Returns**: <code>function</code> - Returns transformed object with snake_case keys  
**See**: [Original method](../actions#actions.rename)  

| Param | Type | Description |
| --- | --- | --- |
| origin | <code>Object</code> | Origin object |
| config | <code>Object</code> | Configuration with struct {field: fn(state)} or {field:value} |

**Example**  
```js
import update from 'rmk/shortcuts/update';

let data = {foo:1}
rename(data, {
   foo: 'bar'
})
// => {bar:1}

let data = [{foo:1}, {foo:2}]
update(data, {
   foo: 'bar'
})
// =>  [{bar:1}, {bar:2}]


let data = {year: 2001}
rename(data, {
    year: (state) => {
      return (state.year.toString().lentgth == 4) ? 'fullYear' : 'year';
    }
})
// => {fullYear: 2001}

let data = [{year: 2001}, {year: 2002}]
rename(data, {
    year: (state) => {
      return (state.year.toString().lentgth == 4) ? 'fullYear' : 'year';
    }
})
// => [{fullYear: 2001}, {fullYear: 2002}]
```
<a name="shortcuts.module.exports"></a>

### shortcuts.module.exports(origin) ⇒ <code>function</code>
Function of converts object keys to camelCase.

**Kind**: static method of <code>[shortcuts](#shortcuts)</code>  
**Returns**: <code>function</code> - Returns transformed object with snake_case keys  
**See**: [Original method](../actions#actions.toCamelCase)  

| Param | Type | Description |
| --- | --- | --- |
| origin | <code>Object</code> | Origin object |

**Example**  
```js
import toCamelCase from 'rmk/shortcuts/toCamelCase';

let data = {FooBar:1, foo_baz:2, bar:3, baz:4}
toCamelCase(data)
// => {fooBar:1, fooBaz:2, bar:3, baz:4}
```
<a name="shortcuts.module.exports"></a>

### shortcuts.module.exports(origin) ⇒ <code>function</code>
Function of converts object keys to snake_case.

**Kind**: static method of <code>[shortcuts](#shortcuts)</code>  
**Returns**: <code>function</code> - Returns transformed object with snake_case keys  
**See**: [Original method](../actions#actions.toCamelCase)  

| Param | Type | Description |
| --- | --- | --- |
| origin | <code>Object</code> | Origin object |

**Example**  
```js
import toCamelCase from 'rmk/shortcuts/toSnakeCase';

let data = {fooBar:1, foo_baz:2, bar:3, Baz:4}
toSnakeCase(data)
// => {foo_bar:1, foo_baz:2, bar:3, baz:4}
```
<a name="shortcuts.module.exports"></a>

### shortcuts.module.exports(origin, config) ⇒ <code>function</code>
Update state by config

**Kind**: static method of <code>[shortcuts](#shortcuts)</code>  
**Returns**: <code>function</code> - Returns transformed object with snake_case keys  
**See**: [Original method](../actions#actions.update)  

| Param | Type | Description |
| --- | --- | --- |
| origin | <code>Object</code> | Origin object |
| config | <code>Object</code> | Configuration with struct {field: fn(state)} or {field:value} |

**Example**  
```js
import update from 'rmk/shortcuts/update';

let data = {foo:1}
update(data, {
   bar: 3
})
// => {foo:1, bar:3}

let data = [{foo:1}, {foo:2}]
update(data, {
   bar: 3
})
// =>  [{foo:1, bar:3}, {foo:2, bar:3}]


let data = {date: new Date(2001)}
update(data, {
    year: (state) => {
      return state.date.getFullYear()
    }
})
// => {date: new Date(2001), year: 2001}

let data = [{date: new Date(2001)}, {date: new Date(2002)}]
update(data, {
    year: (state) => {
      return state.date.getFullYear()
    }
})
// => [{date: new Date(2001), year: 2001}, {date: new Date(2002), year: 2002}]
```
