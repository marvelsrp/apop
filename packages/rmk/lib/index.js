const reduceObject = (origin, remakers) => {
  return remakers.reduce((state, remake) => {
    return remake(state);
  }, origin);
};

const reduceArray = (origin, remakers) => {
  return origin.reduce((state, element) => {
    state.push(reduceObject(element, remakers));
    return state;
  }, []);
};

const reduce = (origin, reamakers) => {
  if (isArray(origin)) {
    return reduceArray(origin, mutators);
  } else if (isObject(origin)) {
    return reduceObject(origin, reamakers);
  } else {
    throw new Error('Wrong input arguments');
  }
};

const rmk = (...reamakers) => {
  return origin => reduce(origin, reamakers);
};

export default rmk;
