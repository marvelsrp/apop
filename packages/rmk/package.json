{
  "name": "@rmk/rmk",
  "version": "1.0.0",
  "scripts": {
    "prepublish": "babel -d ./ ./lib",
    "dev": "babel -d ./ ./lib --watch",
    "test": "jest --coverage",
    "test-dev": "jest --watch"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/tuchk4/rmk.git"
  },
  "dependencies": {
    "@rmk/clear": "=1.0.0"
  },
  "license": "ISC",
  "main": "index.js"
}
