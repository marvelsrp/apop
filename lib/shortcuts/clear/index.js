import rmk from 'rmk';
import rmkActions from 'rmk/actions';

/**
 * Shortcut for clear object
 *
 * @memberof shortcuts
 * @param {Object} origin Origin object
 * @returns {Function} Returns transformed object with snake_case keys
 * @see {@link ../actions#actions.clear Original method}
 * @example
 *
 * import clear from 'rmk/shortcuts/clear';
 *
 * let data = {foo:null, bar: 0, baz:null}
 * clear(data)
 * // => {bar:0}
 */
export default (origin) => {
  return rmk(rmkActions.clear())(origin);
};