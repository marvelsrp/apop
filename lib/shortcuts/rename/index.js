import rmk from '../../';
import rmkActions from '../../actions';

/**
 * Rename state by config
 *
 * @memberof shortcuts
 * @param {Object} origin Origin object
 * @param {Object} config Configuration with struct {field: fn(state)} or {field:value}
 * @returns {Function} Returns transformed object with snake_case keys
 * @see {@link ../actions#actions.rename Original method}
 * @example
 *
 * import update from 'rmk/shortcuts/update';
 *
 * let data = {foo:1}
 * rename(data, {
 *    foo: 'bar'
 * })
 * // => {bar:1}
 *
 * let data = [{foo:1}, {foo:2}]
 * update(data, {
 *    foo: 'bar'
 * })
 * // =>  [{bar:1}, {bar:2}]
 *
 *
 * let data = {year: 2001}
 * rename(data, {
 *     year: (state) => {
 *       return (state.year.toString().lentgth === 4) ? 'fullYear' : 'year';
 *     }
 * })
 * // => {fullYear: 2001}
 *
 * let data = [{year: 2001}, {year: 2002}]
 * rename(data, {
 *     year: (state) => {
 *       return (state.year.toString().lentgth === 4) ? 'fullYear' : 'year';
 *     }
 * })
 * // => [{fullYear: 2001}, {fullYear: 2002}]
 *
 */
export default (origin, config) => rmk(rmkActions.rename(config))(origin);
