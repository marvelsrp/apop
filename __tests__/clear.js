import rmk from '../lib';
import shortcuts from '../lib/shortcuts';

const before = {
  id: 1,
  first_name: 'Stephen',
  last_name: 'Nelson',
  email: 'snelson0@yellowbook.com',
  gender: 'Male',
  ip_address: '74.191.108.58',
  city: 'Kapotnya',
  phrase: 'Universal 6th generation portal',
  null: null,
  undefined,
  date: new Date(2001),
  0: 0,
  1: 1,
  '-1': -1,
  emptyStr: '',
};

test('Clear object', () => {
  const formula = rmk(rmk.clear());

  const result = formula(before);
  expect(result).toMatchSnapshot();
});

test('Clear array of objects', () => {
  const formula = rmk(rmk.clear());

  const result = formula([before]);
  expect(result).toMatchSnapshot();
});

// ---

test('Clear remove undefined and null values', () => {
  const result = shortcuts.clear(before);
  expect(result).toMatchSnapshot();
});

test('Clear array of objects', () => {
  const result = shortcuts.clear([before]);
  expect(result).toMatchSnapshot();
});
