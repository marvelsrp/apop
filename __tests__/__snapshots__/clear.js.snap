// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clear array of objects 1`] = `
Array [
  Object {
    "-1": -1,
    "0": 0,
    "1": 1,
    "city": "Kapotnya",
    "date": 1970-01-01T00:00:02.001Z,
    "email": "snelson0@yellowbook.com",
    "first_name": "Stephen",
    "gender": "Male",
    "id": 1,
    "ip_address": "74.191.108.58",
    "last_name": "Nelson",
    "phrase": "Universal 6th generation portal",
  },
]
`;

exports[`Clear array of objects 2`] = `
Array [
  Object {
    "-1": -1,
    "0": 0,
    "1": 1,
    "city": "Kapotnya",
    "date": 1970-01-01T00:00:02.001Z,
    "email": "snelson0@yellowbook.com",
    "first_name": "Stephen",
    "gender": "Male",
    "id": 1,
    "ip_address": "74.191.108.58",
    "last_name": "Nelson",
    "phrase": "Universal 6th generation portal",
  },
]
`;

exports[`Clear object 1`] = `
Object {
  "-1": -1,
  "0": 0,
  "1": 1,
  "city": "Kapotnya",
  "date": 1970-01-01T00:00:02.001Z,
  "email": "snelson0@yellowbook.com",
  "first_name": "Stephen",
  "gender": "Male",
  "id": 1,
  "ip_address": "74.191.108.58",
  "last_name": "Nelson",
  "phrase": "Universal 6th generation portal",
}
`;

exports[`Clear remove undefined and null values 1`] = `
Object {
  "-1": -1,
  "0": 0,
  "1": 1,
  "city": "Kapotnya",
  "date": 1970-01-01T00:00:02.001Z,
  "email": "snelson0@yellowbook.com",
  "first_name": "Stephen",
  "gender": "Male",
  "id": 1,
  "ip_address": "74.191.108.58",
  "last_name": "Nelson",
  "phrase": "Universal 6th generation portal",
}
`;
